Added a setting to slow down character when moving against a wall

guerro323 opened the pull request:
Based on this feedback from the forum post:
https://forum.unity.com/threads/new-standard-asset-characters-third-person.526612/page-3#post-4045138
I added an option to slow down the character when moving against a wall, by the angle.

---

Variables added in `OpenCharacterController.css`:  

`m_OnWallsKeepVelocity` -> If this option is enabled, the player will not be slowed down when against a wall (as before). If it's disabled, the character will be slowed down.  

`m_OnWallsMaxAngle` -> The maximal angle before the character will not be slowed down:
- If it's at 0°, any angle will be able to slow down the character. 
- If it's at 90°, it will act as if this option was disabled. 
- It's by default at 10° so characters will keep a perfect velocity if it's just touching a bit a wall. 

If some users still want the current behavior, they will just need to enable `m_OnWallsKeepVelocity`.

---

This is my first pull request, so if I missed something in my PR, I'm sorry.
(I think it would be also possible to reduce the code size of this PR)

guerro323 committed:
Updated OCC to add a way to slow down character moving against a wall by the angle.

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -396,6 +396,12 @@ public void OnMoveLoop()
 		bool m_SlideAlongCeiling = true;
 
+		[SerializeField, Tooltip("Should the character keep a perfect velocity when sliding on walls?")]
+		bool m_OnWallsKeepVelocity = true;
```
We prefer to keep features that most users are likely to want as opt-out rather than opt-in, so I suggest you invert this condition.

Try something like `m_SlowAgainstWalls`, and have it be true by default.

hippocoder replied:
I wouldn't have considered slow against walls being default. In fact going by forum, I find people get upset when their characters do not move at a consistent speed, so I advise having this setting disabled by default, which would also work best with people slotting this in as a replacement for the default character controller (which also doesn't slow against walls).

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -396,6 +396,12 @@ public void OnMoveLoop()
+		bool m_OnWallsKeepVelocity = true;
+
+		[SerializeField, Tooltip("The maximal angle to keep a perfect velocity when sliding on walls.")]
+		float m_OnWallsMaxAngle = 10.0f;
```
Similarly to above, I think this field should be inverted in name, and also feels a little awkwardly named to me.

How about `m_MinSlowAgainstWallsAngle`, describing the lowest angle at which we start slowing the character movement?

The current default felt pretty good to me.

Chippit replied:
I also suggest you use the `RangeAttribute` to limit this to the (0, 90) range. The code should also properly endure the 90 case, and treat it exactly the same as if the flag above is false.

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -1836,7 +1842,22 @@ bool CanStickToGround(Vector3 moveVector, out MoveVector getDownVector)
 				// Slide along the obstacle
-				moveVector = project * remainingDistance;
+				// If m_OnWallsKeepVelocity is enabled or the angle is under m_OnWallsMaxAngle, there will be no slowdown
+				var slowDownAngle = Vector3.Angle(project, moveVector);
```
As mentioned in the main comment, a suggested revision:

Chippit replied:
```
float slowDownFactor = 1.0f;

if (!m_OnWallsKeepVelocity)
{
	// Cosine of angle between the movement direction and the tangent is equivalent to the sin of
	// the angle between the movement direction and the normal, which is the sliding component of
	// our movement. 
	var cosine = Vector3.Dot(project, direction);
	var invRescaleFactor = 1 / Mathf.Cos(m_OnWallsMaxAngle * Mathf.Deg2Rad);
	slowDownFactor = Mathf.Clamp01(cosine * invRescaleFactor);
}

moveVector = project * (remainingDistance * slowDownFactor);
```

Chippit replied:
This can be further optimised by precalculating `invRescaleFactor` in Awake() to take out the most expensive part of this calculation. (Probably also want to, in the editor only, recalculate it here so that a user can update the values live in the inspector, but we still get the performance benefit in builds).

Chippit reviewed (changes requested):
Hey, thank you for your contribution! We've tested it and are happy with the results.

I have a few points of technical feedback, however:

Firstly, your linear response curve isn't technically physically accurate. The horizontal component that should determine the speed is geometrically determined by the sin of the angle between the normal and the movement vector. Since we also have the tangent, that's equivalent to the cosine of the angle between that and the movement vector, which is trivial and cheap for us to calculate with a dot product.

Below is an image that compares the two response curves, for your reference.
![response curve](https://user-images.githubusercontent.com/870779/52269486-e355a080-2946-11e9-9d60-577a6c963ac3.PNG)

I have left a suggestion in-line there with code that can achieve the desired response curve, and have also dropped further minor comments in-line in the code.

Secondly, to align with how the inspector works in some other places in the project, the angle setting for this change should be hidden in the inspector if it is not valid based on the state of the bool field. You can refer to `AnimationConfigEditor.cs` for an example of how that is done.

guerro323 committed:
Applied changes from review

guerro323 commented:
Thanks! I updated the file and also added the suggestion for the inspector.
I also remade the way the slow down get called or not (using a single bool variable 'isWallSlowingDown' instead of making a lot of calculations...)

(I don't really like the name of the function `UpdateSlowAgainstWalls` but I didn't found a better name yet)

guerro323 committed:
Forgot to merge the ifs... fixed

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -543,6 +546,8 @@ public void OnMoveLoop()
 		/// </summary>
 		public float defaultHeight { get; private set; }
+		
+		public bool slowAgainstWalls { get { return m_SlowAgainstWalls; } }
```
Good that you added this! We like to ensure all public properties have xmldoc documentation, though, so please add it here just like the other properties.

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -2184,5 +2198,11 @@ void SetRootToOffset()
 		}
+
+		// Update values that are used for computing slow down against walls.
+		void UpdateSlowAgainstWalls()
```
You're right that this name is a bit awkward. I actually just think this method is unnecessary. It's only one line, and only _really_ needs to happen in Awake. I think just inline the code in the two places it's called, and remove this method.

Chippit commented on Assets/_Standard Assets/Characters/Scripts/Physics/OpenCharacterController.cs:
```diff
@@ -593,6 +599,7 @@ void OnValidate()
 			ValidateCapsule(false, ref position);
 			transform.position = position;
 			SetRootToOffset();
+			UpdateSlowAgainstWalls();
```
Good idea putting it here instead of Update, this is actually much better!

Chippit reviewed (changes requested):
Looks great, thanks for the submission. The property without documentation is the only thing that I think is important to update, otherwise this looks good!

guerro323 committed:
Review update

guerro323 committed:
Forgot to remove the comment of the deleted method

guerro323 committed:
comment change

guerro323 commented:
Done!

I just have a small remark, I thought of putting 'slowAgainstWalls' as settable.
So users who are adding the component at runtime can change the behavior.

Should I currently leave it as gettable only? (because if it's going to be settable, I guess there should also be 'slidingDownSlope' property, and others..., which should be for another PR and not for this one)

Chippit commented:
It's a good point, but I think it's okay as it is now. If a user really needs to be able to change it at runtime, perhaps we can look at making that change later (and correctly update the relevant cached values).
